File: sdk/judge.c
#include "bot.h"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
    Usage:
    - ./judge <prog-1> <prog-2> <log-1> <log-2>

    Player protocol:
    - Startup: input "<side>"
        side - 0: first to move, 1: next to move
    - Repeat
        - Update: input "<row> <col>"
            row, col - 0..2
            for the first move, row = col = -1
        - Move: output "<row> <col>"
            row, col - 0..2
*/

static inline char *str_head(char *s, size_t n);

int main(int argc, char *argv[])
{
    /* Initialize players and make sure there are exactly 2 of them */
    int n = bot_player_init(argc, argv);
    if (n != 2) {
        fprintf(stderr, "Expected 2 players, got %d\n", n);
        exit(1);
    }

    /* Inform each player which side it is on */
    bot_player_send(0, "0");
    bot_player_send(1, "1");

    /* Set up board state */
    int move = 0;
    int win = -1, count = 0;
    int row = -1, col = -1;
    int board[3][3];
    memset(board, -1, sizeof board);

    /* Temporary variables and buffers */
    char buf[8];
    char *resp;
    size_t err;

    for (; win == -1 && count < 9; free(resp), move ^= 1) {
        /* Inform the current player of the last move */
        snprintf(buf, sizeof buf, "%d %d", row, col);
        bot_player_send(move, buf);
        resp = bot_player_recv(move, &err, 1000);

        /* An invalid response causes immediate defeat; the same below */
        if (resp == NULL) {
            fprintf(stderr, "Side #%d errors with %d (%s), considered resignation\n",
                move, (int)err, bot_strerr(err));
            win = move ^ 1;
            continue;
        }

        /* Parse the player's response */
        if (sscanf(resp, "%d%d", &row, &col) != 2 ||
            (row < 0 || row >= 3) ||
            (col < 0 || col >= 3))
        {
            fprintf(stderr, "Side #%d format incorrect (%s), considered resignation\n",
                move, str_head(resp, 10));
            win = move ^ 1;
            continue;
        }

        if (board[row][col] != -1) {
            fprintf(stderr, "Side #%d invalid move at (%d, %d), considered resignation\n",
                move, row, col);
            win = move ^ 1;
            continue;
        }

        fprintf(stderr, "Side #%d moves at (%d, %d)\n", move, row, col);
        board[row][col] = move;

        /* Check winning condition */
        for (int i = 0; i < 3; i++)
            if ((board[i][0] == move && board[i][1] == move && board[i][2] == move) ||
                (board[0][i] == move && board[1][i] == move && board[2][i] == move))
            {
                win = move;
                break;
            }
        if ((board[0][0] == move && board[1][1] == move && board[2][2] == move) ||
            (board[0][2] == move && board[1][1] == move && board[2][0] == move))
        {
            win = move;
        }
        if (win != -1) {
            fprintf(stderr, "Side #%d wins!\n", move);
        }
    }

    /* Write the report to stdout */
    printf("{\n  \"winner\": %d,\n  \"board\": \"", win);
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++)
            putchar(".ox"[board[i][j] + 1]);
        printf("\\n");
    }
    printf("\"\n}\n");

    /* Stop all players to ensure their logs are written */
    bot_player_finish();

    return 0;
}

static inline char *str_head(char *s, size_t n)
{
    size_t m = strlen(s);
    if (m > n) {
        if (n >= 3) s[n - 3] = '.';
        if (n >= 2) s[n - 2] = '.';
        if (n >= 1) s[n - 1] = '.';
        s[n] = '\0';
    }
    return s;
}

File: sdk/judge.py
import bot_judge_py as bot_judge
import sys

if bot_judge.init(sys.argv) != 2:
    print('Incorrect number of players!')
    sys.exit()

# Inform each player of their side
bot_judge.send(0, '0')
bot_judge.send(1, '1')

board = [[-1] * 3 for _ in range(3)]
win = -1
row, col = -1, -1   # Last move

for count in range(9):
    move = count % 2    # Current player
    bot_judge.send(move, '%d %d' % (row, col))
    resp, err = bot_judge.recv(move, 1000)

    if resp == None:
        sys.stderr.write(
            'Side #%d errors with %d, considered resignation\n' %
            (move, err))
        win = move ^ 1
        break

    try:
        row, col = map(int, resp.split())
    except:
        row, col = -1, -1

    if row not in range(3) or col not in range(3):
        sys.stderr.write(
            'Side #%d generates invalid response (%s), considered resignation\n' %
            (move, resp if len(resp) < 10 else resp[:7] + '...'))
        win = move ^ 1
        break

    if board[row][col] != -1:
        sys.stderr.write(
            'Side #%d invalid move at (%d, %d), considered resignation\n' %
            (move, row, col))
        win = move ^ 1
        break

    board[row][col] = move

    # Check winning condition
    for i in range(3):
        if ((board[i][0] == move and board[i][1] == move and board[i][2] == move)
         or (board[0][i] == move and board[1][i] == move and board[2][i] == move)):
            win = move
            break

    if ((board[0][0] == move and board[1][1] == move and board[2][2] == move)
     or (board[2][0] == move and board[1][1] == move and board[0][2] == move)):
        win = move

    if win != -1:
        sys.stderr.write('Side #%d wins!\n' % win)
        break

    count += 1

sys.stderr.write('{\n  "winner": %d,\n  "board": "' % win)
sys.stderr.write(
    '\\n'.join(list(
        map(lambda r: ''.join(list(map(lambda x: ".ox"[x + 1], r))), board)
    ))
)
sys.stderr.write('"\n}\n')

bot_judge.finish()
